internal/services/v1/schema.go

```diff
@@ -88,7 +88,10 @@ func (ss *schemaServer) WriteSchema(ctx context.Context, in *v1.WriteSchemaReque
 	// Do as much validation as we can before talking to the datastore
 	newDefs := strset.NewWithSize(len(nsdefs))
 	for _, nsdef := range nsdefs {
-		ts, err := namespace.BuildNamespaceTypeSystemForDefs(nsdef, nsdefs)
+		ts, err := namespace.NewNamespaceTypeSystem(nsdef,
+			namespace.ResolverForPredefinedDefinitions(namespace.PredefinedElements{
+				Namespaces: nsdefs,
+			}))
```

Comment:
> If I understand correctly, we are basically turning the `core.Namespace` into a type system for validation purposes. Are we intentionally not including the caveats here because it will be done in a follow up PR?

The type system was already in the namespace package; you are correct that the caveat-specific stuff gets added in the followup PR, but I'm also going to be breaking out the type system stuff into its own package in that PR (or a followup) as well

> and also for my own understanding, is the call below to `AnnotateNamespace` just done for validation purposes and we don't actually need to annotate it right now?

No; its explicitly done to annotate the namespace with inferred data (such as canonical cache keys) before it is written in the namespace store

Reply:
> If I understand correctly, we are basically turning the `core.Namespace` into a type system for validation purposes. Are we intentionally not including the caveats here because it will be done in a follow up PR?

The type system was already in the namespace package; you are correct that the caveat-specific stuff gets added in the followup PR, but I'm also going to be breaking out the type system stuff into its own package in that PR (or a followup) as well

> and also for my own understanding, is the call below to `AnnotateNamespace` just done for validation purposes and we don't actually need to annotate it right now?

No; its explicitly done to annotate the namespace with inferred data (such as canonical cache keys) before it is written in the namespace store